engine/src/Audio/Audio.fs

```diff
@@ -156,31 +157,32 @@ module Song =
     let set_global_offset (offset) = _global_offset <- offset
 
     let private song_loader =
-        { new Async.SwitchService<string option * SongLoadAction * bool, Song * SongLoadAction * bool>() with
-            override this.Process((path, after_load, song_playing)) =
+        { new Async.SwitchService<string option * SongLoadAction, Song * SongLoadAction>() with
+            override this.Process((path, after_load)) =
                 async {
                     return
                         match path with
-                        | Some p -> Song.FromFile p, after_load, song_playing
-                        | None -> Song.Default, after_load, song_playing
+                        | Some p -> Song.FromFile p, after_load
+                        | None -> Song.Default, after_load
                 }
 
-            override this.Handle((song, after_load: SongLoadAction, song_playing: bool)) =
+            override this.Handle((song, after_load: SongLoadAction)) =
                 loading <- false
                 now_playing <- song
                 change_rate rate
 
                 match after_load with
                 | SongLoadAction.PlayFromPreview ->
-                    (if song_playing then play_from else seek) preview_point
+                    (if song_was_playing then play_from else seek) preview_point
                 | SongLoadAction.PlayFromBeginning ->
-                    (if song_playing then play_from else seek) 0.0f<ms>
+                    (if song_was_playing then play_from else seek) 0.0f<ms>
                 | SongLoadAction.Wait -> ()
+                song_was_playing <- false
         }
 
     let change (path: string option, offset: Time, new_rate: float32, (preview: Time, chart_last_note: Time), after_load: SongLoadAction) =
-        let song_was_playing = playing() || load_path = None
         let path_changed = path <> load_path
+        if path_changed then song_was_playing <- song_was_playing || playing() || load_path = None
```

Comment:
I specifically `if path_changed then` which means song_was_playing can only become initially true if song_loader is about to receive a request (and therefore set song_was_playing to false when a song was playing on first successful load)
song_was_playing feeds into itself since if we try to load a song while already loading another song, every available indicator (loading, IsRunning) isn't enough (as far as I can tell) to tell if the first song queued should start playing on load

Reply:
Yep makes sense to me now, thanks a lot for finding this

I've pushed an alternate version that should do the same thing, but I think is a bit clearer to read later on